src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py

```diff
@@ -0,0 +1,121 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Formula generator from component graph for Consumer Power."""
+
+from __future__ import annotations
+
+from collections import abc
+
+from ....microgrid import connection_manager
+from ....microgrid.component import Component, ComponentCategory, ComponentMetricId
+from .._formula_engine import FormulaEngine
+from .._resampled_formula_builder import ResampledFormulaBuilder
+from ._formula_generator import ComponentNotFound, FormulaGenerator
+
+
+class ConsumerPowerFormula(FormulaGenerator):
+    """Formula generator from component graph for calculating the Consumer Power."""
+
+    def generate(self) -> FormulaEngine:
+        """Generate formula for calculating consumer power from the component graph.
+
+        Returns:
+            A formula engine that will calculate the consumer power.
+
+        Raises:
+            ComponentNotFound: If the component graph does not contain a consumer power
+                component.
+        """
+        builder = self._get_builder("consumer-power", ComponentMetricId.ACTIVE_POWER)
+        component_graph = connection_manager.get().component_graph
+        grid_component = next(
+            (
+                comp
+                for comp in component_graph.components()
+                if comp.category == ComponentCategory.GRID
+            ),
+            None,
+        )
+
+        if grid_component is None:
+            raise ComponentNotFound("Grid component not found in the component graph.")
+
+        grid_successors = component_graph.successors(grid_component.component_id)
+        if not grid_successors:
+            raise ComponentNotFound("No components found in the component graph.")
+
+        if len(grid_successors) == 1:
+            grid_meter = next(iter(grid_successors))
+            assert (
+                grid_meter.category == ComponentCategory.METER
+            ), "Only grid successor is not a meter."
+            return self._gen_with_grid_meter(builder, grid_meter)
+        return self._gen_without_grid_meter(builder, grid_successors)
+
+    def _gen_with_grid_meter(
```

Comment:
Are we going to introduce sub panels into the component graph?  I haven't heard of any such plans.   Those are passive components right?  i.e we can't read from them or control them.

Reply:
As far as I remember that was the plan. @tiyash-basu-frequenz can you confirm?

Reply:
I suggest to add a short warning comment to the documentation such that we are aware in the future.

Reply:
I think the term "sub panels" is misleading. If I have understood it correctly, then it is a section of the microgrid behind a dedicated fuse. I'd call it sub-section.

But I'd suggest ignoring it for now. It is not scheduled to be added to the microgrid API soon. To me even a comment seems unnecessary: because it does not improve the readability and just hints at a future functionality without knowing when we will add it. But up to you.

Reply:
If you really want to track this feature, then creating an issue is the proper way to do it.

Reply:
You are right w.r.t. the issue. We should create one but I'd also leave a comment just to have the behavior documented because it tells that the algorithm isn't going into the depth.

Reply:
Then I'd suggest to rather create the issue and add a link to the issue in the comment.

I'd suggest doing it in moderation, if never at all, because you do not want to end up adding every proposed feature in the comments.